course04/homework00/matteotonelli/main.cc

```diff
@@ -0,0 +1,144 @@
+#include <iostream>
+#include <array>
+#include <vector>
+#include <algorithm>
+
+class OrderedArray
+{
+public:
+    static constexpr const std::size_t MaxSize = 1000;
+
+    OrderedArray();
+    OrderedArray(const std::vector<int>&);
+
+    std::size_t size() const { return mSize; }
+    std::size_t max_size() const { return MaxSize; }
+    std::size_t capacity() const { return MaxSize; }
+
+    void push_back(int);
+
+    const int* c_array() const { return mData.data(); }
+
+private:
+    std::size_t mSize = 0;
+    std::array<int, MaxSize> mData;
+};
+
+OrderedArray::OrderedArray()=default;
+
+OrderedArray::OrderedArray(const std::vector<int>& v)
+{
+    if (v.size() <= MaxSize)
+    {
+        mSize = v.size();
+    }
+    else{
+        throw std::out_of_range("Fermo!!! diocan!");
+    }
+
+    std::copy(v.data(), v.data() + mSize, mData.data());
+    std::sort(mData.data(), mData.data() + mSize);
+}
+
+void OrderedArray::push_back(int newElement)
+{
+    if (mSize >= MaxSize)
+        throw std::out_of_range("...");
+
+    size_t i = 0;
+    size_t j = mSize;
+    while ((i < mSize) && (mData[i] < newElement))
+        i++;
+
+    while (j > i)
+    {
+        mData[j] = mData[j-1];
+        --j;
+    }
```

Comment:
You shouldnt do all this yourself: its complicated and STL algorithms are faster